main.cc

```diff
@@ -544,6 +544,85 @@ static locator::host_id initialize_local_info(sharded<db::system_keyspace>& sys_
     return linfo.host_id;
 }
 
+static bool raft_topology_change_enabled(const db::config& cfg) {
+    return cfg.consistent_cluster_management() &&
+        cfg.check_experimental(db::experimental_features_t::feature::CONSISTENT_TOPOLOGY_CHANGES);
+}
+
+static void initialize_system_keyspace(sharded<db::system_keyspace>& sys_ks,
+        sharded<cql3::query_processor>& qp,
+        sharded<locator::snitch_ptr>& snitch,
+        distributed<replica::database>& db,
+        sharded<locator::effective_replication_map_factory>& erm_factory,
+        const db::config& cfg,
+        sharded<gms::feature_service>& feature_service)
+{
+    supervisor::notify("starting system keyspace");
+    sys_ks.start(std::ref(qp), std::ref(db), std::ref(snitch)).get();
+    // TODO: stop()?
+
+    // Initialization of a keyspace is done by shard 0 only. For system
+    // keyspace, the procedure  will go through the hardcoded column
+    // families, and in each of them, it will load the sstables for all
+    // shards using distributed database object.
+    // Iteration through column family directory for sstable loading is
+    // done only by shard 0, so we'll no longer face race conditions as
+    // described here: https://github.com/scylladb/scylla/issues/1014
+    supervisor::notify("loading system sstables");
+    replica::distributed_loader::init_system_keyspace(sys_ks, erm_factory, db, system_table_load_phase::phase1).get();
+
+    // Re-enable previously enabled features on node startup.
+    // This should be done before commitlog starts replaying
+    // since some features affect storage.
+    feature_service.local().enable_features_on_startup(sys_ks.local(), raft_topology_change_enabled(cfg)).get();
+
+    db.local().maybe_init_schema_commitlog();
+
+    // Init schema tables only after enable_features_on_startup()
+    // because table construction consults enabled features.
+    // Needs to be before system_keyspace::setup_after_commitlog(), which writes to schema tables.
+    replica::distributed_loader::init_system_keyspace(sys_ks, erm_factory, db, system_table_load_phase::phase2).get();
+    sys_ks.local().setup_before_commitlog().get();
+
+    supervisor::notify("starting schema commit log");
+
+    // Check there is no truncation record for schema tables.
+    // Needs to happen before replaying the schema commitlog, which interprets
+    // replay position in the truncation record.
+    // Needs to happen before system_keyspace::setup(), which reads truncation records.
+    db.local().get_tables_metadata().for_each_table([] (table_id, lw_shared_ptr<replica::table> table_ptr) {
```

Comment:
I think this code should be removed or hidden before an upgrade switch. It should be possible to truncate system tables.

Reply:
@tgrabiec

Reply:
Schema tables cannot be safely truncated at run-time, because they must match in-memory state. If you want to do that, you have to stop the node and delete the sstables.

Reply:
It is not true for just any schema table @tgrabiec 
I think people who truncate schema tables at runtime know what they are doing.

Reply:
Any meddling with schema tables by a user is not allowed. By not allowing ourselves truncate we just shoot ourselves in the foot.

Reply:
You mean it should be allowed internally, but disabled for the user?

What's the use case?

Anyhow, the system is really not prepared for supporting that. The node will crash afterwards. So if we want to support that, it takes much more work than just removing this check.

Reply:
The proper way to achieve clean state of schema tables is to issue DROP statements.

Reply:
Can you come up with a case when e.g. view build status or column mapping history tables will get broken if truncated? These are not "schema" tables, perhaps, but system tables. I don't have a use case. I can't imagine a user trying to truncate system tables/schema tables, but now discussing it with you perhaps I think it is a bad idea even for our own staff. However, I think the guardrails for this are redundant.

Reply:
The guardrails we have are specifically for system_schema.* tables. Other system tables may be fine to truncate.